_drafts/javascript.md

```diff
@@ -0,0 +1,463 @@
+
+> Modernize the project by translating CoffeeScript to JavaScript, improving integration with various developer tools, and adding features that make it easier for developers to automate their workflow.
+
+**Scope:**
+
+- Simplify contribution / maintenance
+- Removal of tooling / build steps
+- Align Node support with [Node’s LTS schedule](https://github.com/nodejs/LTS#lts-schedule1)
+- maintain backwards compatibility of user scripts written in CoffeeScript
+
+**Out of scope:**
+
+- no new features or breaking changes (beyond dropping support for Node <v4)
+- Introduce any other tooling besides linter
+- Any non-trivial refactoring of the existing code base
+
+## Proposed solution
+
+Converting of all CoffeeScript to JavaScript which can run natively on Node 4 ([current maintenance version until April 2018](https://github.com/nodejs/LTS#lts-schedule1)).
+
+Probably the biggest challenge is that CoffeeScript classes are incompatible
+with es2015 classes: they cannot extend from each other. It is something that
+[CoffeeScript v2 is addressing](http://coffeescript.org/v2/#coffeescript-2).
+
+The hubot eco system expects CoffeeScript-compatible "classes" being return in
+`require('hubot')`, even when 3rd party adapters have been implemented in
+JavaScript themselves. To address that problem, I suggest we use es2015 `class`
+declarations internally but wrap them in CoffeeScript compatible classes in
+`/index.js`. Additionally, we can introduce a new officially support export at
+`require('hubot/es2015')` which returns our es2015 class declarations, so that
+3rd party developers can use es2015 class declarations themselves and inhert
+from ours.
+
+## Detailed process
+
+The steps of the conversation will be
+
+- [x] Convert source files from CoffeeScript to JavaScript.
+- [x] Test with popular projects depending on Hubot
+- [x] Convert test files from CoffeeScript to JavaScript.
+- [x] update package.json
+- [x] add script for JavaScript linting
+- [ ] Release `hubot@3.0.0`
+  - [ ] Update hubot’s version range `peerDependencies` of default scripts
+  - [ ] Publish `hubot@3.0.0` with `--tag=next`
+  - [ ] Convert default scripts to JavaScrip, release as new breaking versions with `@next` tag
+  - [ ] Convert generator to JavaScript, release new breaking version with `@next` tag
+  - [ ] Test all `@next` releases
+  - [ ] Set `@latest` tag on all modules
+- [ ] Update documentation
+
+All steps will be part of the same PR but split into separate commits so people
+can follow more easily.
+
+See also [examples](#examples) below
+
+### Convert source files from CoffeeScript to JavaScript
+
+Convert all source files from CoffeeScript to JavaScript with a tool like [decaffeinate](https://github.com/decaffeinate/decaffeinate), see also [github/hubot#1138](https://github.com/github/hubot/issues/1138).
+
+Make sure the existing tests (still written in CoffeeScript) run against the new JavaScript.
```

Comment:
My previous hubot -> ES6 porting experience indicate `porting tests first` works well. Not sure for the reverse route.